_posts/uportal/2017-12-06-myuw-fork.md

```diff
@@ -0,0 +1,168 @@
+summary:    MyUW has customized uPortal locally in various ways. What's worth saying about this?
+tags:       [uPortal]
+---
+
+# Summary
+
+MyUW local uPortal is customized ("hacked" or "forked") from baseline uPortal 
+("upstream") in various ways.
+
+Modifications include
+
++ **Rendering pipeline branching** to direct some traffic out of the 
+traditional rendering pipeline for handling via `uPortal-home`. The traditional 
+rendering pipeline only renders maximized portlets.
++ New **uPortal theme and skins** that make uPortal look like `uPortal-home` 
+when rendering a maximized portlet.
++ **Additional JSON APIs** and additional data in baseline uPortal JSON APIs
++ Layout management modifications to default adding a portlet to the first 
+rather than last position in a container. (This makes the "Add to home" from 
+`uPortal-home` a better experience.)
++ Removing features MyUW doesn't use. Most portlets. Related apps. Portlet 
+events.
++ Local data.
+
+These customizations are interesting because
+
++ They amount to hints about what another adopter might need to do to get 
+`uPortal-home` locally working.
++ They're potentially valuable to port forward into upstream.
+
+
+# Why is MyUW uPortal forked?
+
+MyUW is running uPortal 4.2.1 with local customizations that both pre-date and post-date that release, the former achieved through heroic source control merging of a long history of local change.
+
+This was closer to a typical historical uPortal adoption model, building the 
+local portal from source. With ready access to that source, it's tempting to 
+enhance it. With long release cycles and difficult upgrades, making 
+enhancements directly in upstream often felt like duplicated effort, since to 
+realize the improvement locally one also had to make it locally.
+
+A more modern adoption model would obsessively make these improvements in 
+upstream rather than locally and then locally adopt them by upgrading to open 
+source product releases with the improvements.
+
+# The modifications
+
+## Hacked rendering pipeline
+
+MyUW configures the uPortal rendering pipeline to direct to 
+`uPortal-home` the requests that can be handled by `uPortal home`. The 
+traditional rendering pipeline still renders portlets in maximized and 
+exclusive window states.
+
++ Prefer to render static content via `/web/static/{fname}`
++ Prefer to render portlets via `/web/exclusive/{fname}` (`renderOnWeb`)
++ Bounce attempts to do anything other than render just one portlet (i.e. maximized or exclusive) to `/web`
++ Coerce guest user to the desired profile
+
+# Hacked theme
+
+MyUW uses a customized "Bucky" theme forked from Respondr. The purpose of this customized theme and associated skins is to make rendering a portlet maximized via uPortal look like rendering uPortal-home, so that users switch reasonably seamlessly between the two approaches depending on content.
+
++ Fewer regions. Simpler. Remove features MyUW isn't using.
```

Comment:
uPortal 5 has changed a number of regions to portlets that can easily be removed. E.G. https://github.com/Jasig/uPortal/pull/1081